### PR TITLE
Projected gradient descent

### PR DESCRIPTION
**Pull request by grishabh147**
Implemented a projected gradient descent-based solver as an illustrative example to show how a subproblem solver can be interfaced with `BlockNLPAlgorithms.jl`.

### CODE REVIEW

File: test/my_solver.jl [hunk 1]
```diff
@@ -0,0 +1,62 @@
+struct MySolver <: AbstractBlockSolver
+    MySolver() = begin
```
Comment (frapac) [inline test/my_solver.jl]: I would prefer writing explicitly: 
```julia
function MySolver()
    ...
end
```

File: test/my_solver.jl [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        @warn "This solver does not provide dual variable information."
+        new()
+    end
+end
+
+mutable struct PGDSolverInstance
```
Comment (frapac) [inline test/my_solver.jl:8]: Does it have to be a `mutable struct`? I would prefer using just `struct` if possible

File: test/runtests.jl [hunk 3]
```diff
@@ -91,9 +93,8 @@ function BlockNLPAlgorithms.initialize_solver(
     nlp_blocks::Vector{<:AbstractNLPModel},
 )
     nb = length(nlp_blocks)
-    ips = Vector{MadNLP.InteriorPointSolver}(undef, nb)
+    ips = [MadNLP.InteriorPointSolver(nlp_blocks[i]; solver.options...) for i in 1:nb]
```
Comment (frapac) [inline test/runtests.jl:96]: Better!

**Review by frapac — COMMENTED** [review]
Looks good to me, I just have two minor remarks but ok to merge once they are addressed.

File: test/my_solver.jl (hunk 2 above)
Comment (grishabh147) [reply test/my_solver.jl:8]: In this case, we need it to be a `mutable struct.` I am using this to initialize some arrays that are updated in every iteration of the gradient descent algorithm.

Comment (frapac) [reply test/my_solver.jl:8]: Ok then!

**Review by frapac — APPROVED** [review]

**Pull request merged**